scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -160,41 +188,112 @@
+    ;
+    ;   Note that we have release semantics. We could fail to process any operation. This means that, where relevant, we
+    ;     first write to datomic, then change into the next step. We thus show some intermediate states in the machine
+    ;     that seem like they could be skipped.
+    ;
+    ;   In the (:completed,*) states, we delete from datomic, transitioning into (:completed,:missing) and thence to (:missing, :missing)
+    ;     When we're in a terminal pod state.
+    ;
+    ;   We only delete a pod if and only if the pod is in a terminal (or unknown) state.
+    ; In some cases, we may mark mea culpa retries (if e.g., the node got reclaimed, so the pod got killed). In weird cases, we always need to mark mea culpa retries.
     (let
       [new-expected-state-dict (case (vector (or expected-state :missing) (or (:state synthesized-state) :missing))
                                  [:expected/starting :missing] (launch-task api-client expected-state-dict)
                                  [:expected/starting :pod/running] (pod-has-started kcc existing-state-dict)
-                                 ; TODO We need to add cases for the other [:expected/starting *] states in.
-                                 [:expected/starting :pod/waiting] expected-state-dict
-                                 [:expected/starting :pod/succeeded] (pod-has-just-completed kcc existing-state-dict)
-                                 [:expected/starting :pod/failed] (pod-has-just-completed kcc existing-state-dict)
+                                 [:expected/starting :pod/waiting] expected-state-dict ; Its starting. Can be stuck here.
+                                 [:expected/starting :pod/succeeded] (pod-has-just-completed kcc existing-state-dict) ; Finished fast.
+                                 [:expected/starting :pod/failed] (pod-has-just-completed kcc existing-state-dict) ; TODO: May need to mark mea culpa retry
                                  [:expected/running :pod/running] expected-state-dict
                                  [:expected/running :pod/succeeded] (pod-has-just-completed kcc existing-state-dict)
-                                 [:expected/completed :pod/succeeded] (remove-finalization-if-set-and-delete api-client expected-state-dict pod)
-                                 [:expected/running :pod/failed] (pod-has-just-completed kcc existing-state-dict)
-                                 [:expected/completed :pod/failed] (remove-finalization-if-set-and-delete api-client expected-state-dict pod)
+                                 [:expected/completed :pod/succeeded]
+                                 ; This is an exception. The writeback to datomic has occurred, so there's nothing to do except to delete the task from kubernetes
+                                 ; and remove it from our tracking.
+                                 (delete-task api-client pod)
```

Comment:
For consistency, please begin next to the vector on the same line. The comments can be placed above the line.